httpbp/trust_handers.go

```diff
@@ -0,0 +1,220 @@
+	"github.com/reddit/baseplate.go/signing"
+)
+
+const (
+	edgeContextSignatureHTTPHeader = "X-Edge-Request-Signature"
+	spanSignatureHTTPHeader        = "X-Span-Signature"
+)
+
+// HeaderTrustHandler provides an interface PopulateBaseplateRequestContext to
+// verify that it should trust the HTTP headers it receives.
+type HeaderTrustHandler interface {
+	// TrustEdgeContext informs the function returned by PopulateBaseplateRequestContext
+	// if it can trust the HTTP headers that can be used to create an edge
+	// context.
+	//
+	// If it can trust those headers, then the headers will be copied into the
+	// context object to be later used to initialize the edge context for the
+	// request.
+	TrustEdgeContext(r *http.Request) bool
+
+	// TrustSpan informs the function returned by PopulateBaseplateRequestContext
+	// if it can trust the HTTP headers that can be used to create a server
+	// span.
+	//
+	// If it can trust those headers, then the headers will be copied into the
+	// context object to later be used to initialize the server span for the
+	// request.
+	TrustSpan(r *http.Request) bool
+}
+
+// NeverTrustHeaders implements the HeaderTrustHandler interface and always
+// returns false.
+type NeverTrustHeaders struct{}
+
+// TrustEdgeContext always returns false.  The edge context headers will never
+// be added to the context.
+func (h NeverTrustHeaders) TrustEdgeContext(r *http.Request) bool {
+	return false
+}
+
+// TrustSpan always returns false.  The span headers will never be added to the
+// context.
+func (h NeverTrustHeaders) TrustSpan(r *http.Request) bool {
+	return false
+}
+
+// AlwaysTrustHeaders implements the HeaderTrustHandler interface and always
+// returns true.
+type AlwaysTrustHeaders struct{}
+
+// TrustEdgeContext always returns true.  The edge context headers will always
+// be added to the context.
+func (h AlwaysTrustHeaders) TrustEdgeContext(r *http.Request) bool {
+	return true
+}
+
+// TrustSpan always returns true.  The span headers will always be added to the
+// context.
+func (h AlwaysTrustHeaders) TrustSpan(r *http.Request) bool {
+	return true
+}
+
+// HMACTrustHandler implements the HeaderTrustHandler interface and checks the
+// headers for a valid signature header.  If the headers are signed, then they
+// can be trusted and the Trust request returns true.  If there is no signature
+// or the signature is invalid, then the Trust request returns false.
+//
+// For both the span and edge context headers, the trust handler expects the
+// caller to provide the signature of a message in the following format:
+//
+// 		"{header0}:{value0}|{header1}|{value1}|...|{headerN}:{valueN}"
+//
+// where the headers are sorted lexicographically.  Additionally, the signature
+// should be generated using the baseplate provided `signing.Sign` function.
+//
+// HMACTrustHandler provides implementations for both signing and verifying
+// edge context and span headers.
+type HMACTrustHandler struct {
```

Comment:
I wonder if we should rename this to something like `BPSignatureTrustHandler` since the handler itself doesn't know or care about the signing method used?

Reply:
Or maybe `HeaderSignatureTrustHandler`